src/Razor/src/Microsoft.VisualStudio.LanguageServices.Razor/ProjectSystem/WindowsRazorProjectHostBase.cs

```diff
@@ -104,59 +104,166 @@ internal Task LoadAsync()
+        // important that we maintain our list of subscriptions because if a slice is removed, we are responsible for cleaning
+        // up our resources.
+        //
+        // It's worth noting the events also give us a key, which is a list of "dimensions". We only care about the target framework
+        // but they could be strictly anything, and could change at any time. If they do, we'll get new events, so the easiest
+        // thing to do is just treat the key as an opaque object. CPS implements IEquatable on it, expressly for this purpose.
+        // We should not have any logic that depends on the contents of the key.
+        //
+        // Somewhat similar to https://github.com/dotnet/project-system/blob/bf4f33ec1843551eb775f73cff515a939aa2f629/src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Tree/Dependencies/Subscriptions/DependenciesSnapshotProvider.cs
+        // but a lot simpler.
+        _disposables.Add(CommonServices.ActiveConfigurationGroupSubscriptionService.SourceBlock.LinkTo(
+            DataflowBlockSlim.CreateActionBlock<IProjectVersionedValue<ConfigurationSubscriptionSources>>(SlicesChanged, nameFormat: "Slice {1}"),
+            new DataflowLinkOptions() { PropagateCompletion = true }));
+
+        // Join, in the JTF sense, the ActiveConfigurationGroupSubscriptionService, to help avoid hangs in our OnProjectChangedAsync method
+        _disposables.Add(ProjectDataSources.JoinUpstreamDataSources(CommonServices.ThreadingService.JoinableTaskFactory, CommonServices.FaultHandlerService, CommonServices.ActiveConfigurationGroupSubscriptionService));
+
         return Task.CompletedTask;
     }
 
+    private void SlicesChanged(IProjectVersionedValue<ConfigurationSubscriptionSources> value)
+    {
+        // Create a new dictionary representing the subscriptions we know about at the start of the update. Data flow ensures
+        // this method will not be called in parallel.
+        var current = new Dictionary<ProjectConfigurationSlice, IDisposable>(_projectSubscriptions);
+
+        foreach (var (slice, source) in value.Value)
+        {
+            if (!_projectSubscriptions.TryGetValue(slice, out var dataSource))
+            {
+                Assumes.False(current.ContainsKey(slice));
+
+                // This is a new slice that we didn't previously know about, either because its a new target framework, or how dimensions
+                // are calculated has changed. We simply subscribe to updates for it, and let our action block code handle whether the
+                // distinction is important. To put it another way, we may end up having multiple subscriptions and events that would be
+                // affect about the same project.razor.json file, but our event handling code ensures we don't handle them more than
+                // necessary.
+                var subscription = source.JointRuleSource.SourceBlock.LinkTo(
```

Comment:
When subscribing to JointRule source, you're really subscribing to two sources internally (ProjectRule for evaluation and BuildRule for dtb) and getting SyncLink behaviour between them.

What that means is that CPS needs to take your rule names here and divide them in two, so it subscribes correctly to each of those upstream sources.

You can help CPS avoid that work by passing a `JointRuleDataflowLinkOptions` instance for the `DataflowLinkOptions` object, and calling a different overload of `LinkTo`.

Also internally that type will build an `ImmutableHashSet<string>` for the rule names. You can avoid that overhead by having your `GetRuleNames` return `ImmutableHashSet<string>` itself (or `GetEvaluationRuleNames` and `GetBuildRuleNames` if you do use `JointRuleDataflowLinkOptions`).

Reply:
Happy to change GetRuleNames. I'm not so convinced about breaking them up into evaluation and build rules, as I don't think we on the Razor team know which is which. I certainly don't :)

Reply:
Any rule that has a data source specifying a target name is from build. Otherwise it's from evaluation. It should be quite straightforward to decide for each one.

Reply:
Is there a way to tell without finding the .xaml file for the rule?

This is all making me think we should just use Roslyn events, then you folks will be responsible for the CPS code 😛

Reply:
I can't find the XAML rule files in this repo. The generated partial classes are there. The project file says it adds the XAML files as None items. But they don't appear to be in the repo. I'm very confused.

Reply:
Oh, they're coming from a NuGet package: https://github.com/dotnet/razor/blob/ac7208c88a08bd273b4f9b6c84788eefc5dd70b8/src/Razor/src/Microsoft.VisualStudio.LanguageServices.Razor/Microsoft.VisualStudio.LanguageServices.Razor.csproj#L7

Reply:
Digging into that package in NuGet package explorer.

From build:

- `RazorComponentWithTargetPath`
- `RazorGenerateWithTargetPath`

From evaluation:

- `RazorConfiguration`
- `RazorExtension`
- `RazorGeneral`

Reply:
Its the [other three](https://github.com/dotnet/razor/blob/main/src/Razor/src/Microsoft.VisualStudio.LanguageServices.Razor/ProjectSystem/FallbackWindowsRazorProjectHost.cs#L35-L37) I was mainly wondering about. eg, is `Compile.xaml` in the project-system repo? CPS? or is it a copy of `None.xaml` that CPS makes internally, like when you use `AvailableItemName`? The answers to those questions seem like they would be dark magic to a future razor tooling developer, for what I'm guessing is a pretty minor perf improvement.

Reply:
That's fair. Also if the rules changed source over time, it'd break the code. The perf diff would be minor, but it's during project load which is a sensitive time for perf. I don't have any issue with you leaving this as-is.